[qca-nss-ecm] Do not process bridged packets twice
Bridged packets go through both bridge and IP post
routing hook. We should detect the packet in the IP
post routing hook and skip this process.

Signed-off-by: Murat Sezgin <msezgin@codeaurora.org>
Change-Id: Icb179e72d0ac572fe3a8c0a1a0fe1b6f553792c3

diff --git a/frontends/nss/ecm_nss_ported_ipv4.c b/frontends/nss/ecm_nss_ported_ipv4.c
index e94da79..23b3608 100644
--- a/frontends/nss/ecm_nss_ported_ipv4.c
+++ b/frontends/nss/ecm_nss_ported_ipv4.c
@@ -2483,6 +2483,20 @@ done:
 		;
 	}
 
+	/*
+	 * Bridged traffic goes through the IP post routing hook as well after it
+	 * finishes the bridge post routing hook. In that case, ecm_dir will become
+	 * as Non-Nat since the is_routed flag is true. But the is_routed flag of the connection
+	 * was set as false while the packet was going throught the bridge post routing
+	 * hook. So, we need to check if the is_routed flag and ecm_dir matches the routed
+	 * flow condition. If it doesn't, do not process the flow.
+	 */
+	if (!ecm_db_connection_is_routed_get(ci) && (ecm_dir == ECM_DB_DIRECTION_NON_NAT)) {
+		DEBUG_TRACE("%px: ignore route hook path for bridged packet\n", ci);
+		ecm_db_connection_deref(ci);
+		return NF_ACCEPT;
+	}
+
 #ifdef CONFIG_NET_CLS_ACT
 	/*
 	 * Check if IGS feature is enabled or not.
diff --git a/frontends/nss/ecm_nss_ported_ipv6.c b/frontends/nss/ecm_nss_ported_ipv6.c
index aede82f..95ea7c5 100644
--- a/frontends/nss/ecm_nss_ported_ipv6.c
+++ b/frontends/nss/ecm_nss_ported_ipv6.c
@@ -2299,6 +2299,20 @@ done:
 		;
 	}
 
+	/*
+	 * Bridged traffic goes through the IP post routing hook as well after it
+	 * finishes the bridge post routing hook. In that case, ecm_dir will become
+	 * as Non-Nat since the is_routed flag is true. But the is_routed flag of the connection
+	 * was set as false while the packet was going throught the bridge post routing
+	 * hook. So, we need to check if the is_routed flag and ecm_dir matches the routed
+	 * flow condition. If it doesn't, do not process the flow.
+	 */
+	if (!ecm_db_connection_is_routed_get(ci) && (ecm_dir == ECM_DB_DIRECTION_NON_NAT)) {
+		DEBUG_TRACE("%px: ignore route hook path for bridged packet\n", ci);
+		ecm_db_connection_deref(ci);
+		return NF_ACCEPT;
+	}
+
 #ifdef CONFIG_NET_CLS_ACT
 	/*
 	 * Check if IGS feature is enabled or not.
